Improvement/zenko 4499 iam policy common

KazToozs opened the pull request:
**What does this PR do, and why do we need it?**
Adding functionality for iam policy testing with CTST.

**Which issue does this PR fix?**
Part of https://scality.atlassian.net/jira/software/c/projects/OS/boards/214?modal=detail&selectedIssue=ZENKO-4480&assignee=61360619c425a20068e9dfcb&assignee=602aa4981bf1ce006b9de203&assignee=6113892de6e6f8007168633b


**Special notes for your reviewers**:
This PR relies on CTST v0.2.2 which is not yet released and will fail at first.

bert-e commented:
# Hello kaztoozs,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/zenko/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-4499 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.6.10`


Please check the `Fix Version/s` of ZENKO-4499, or the target
branch of this pull request.

XinLi0207 commented on .github/scripts/end2end/run-e2e-ctst.sh:
```diff
@@ -36,4 +53,5 @@ kubectl run $POD_NAME \
         --env=TARGET_VERSION=$VERSION  \
-        -- ./run "$COMMAND" $WORLD_PARAMETERS "--parallel $PARALLEL_RUNS --retry $RETRIES --retry-tag-filter @Flaky"
+        --env=VERBOSE=1 \
+        -- ./run "$COMMAND" $WORLD_PARAMETERS "--parallel $PARALLEL_RUNS --retry $RETRIES --retry-tag-filter @Flaky --tags @IAM-Policies-ARWWI"
```
let's remove this tag  `--tags @IAM-Policies-ARWWI"`

XinLi0207 commented on .github/workflows/end2end.yaml:
```diff
@@ -634,4 +631,4 @@ jobs:
           user: ${{ secrets.ARTIFACTS_USER }}
           password: ${{ secrets.ARTIFACTS_PASSWORD }}
-          source: artifacts
+          source: artifacts
```
let's bring back the blank line

XinLi0207 commented on tests/ctst/Dockerfile:
```diff
@@ -1,12 +1,19 @@
 ARG CTST_TAG
-FROM registry.scality.com/cli-testing/cli-testing:$CTST_TAG
+FROM registry.scality.com/cli-testing-dev/cli-testing:$CTST_TAG
```
leave a note here, so that we don't forget to change here once ctst is released

XinLi0207 commented on tests/ctst/common/common.ts:
```diff
@@ -15,3 +17,33 @@ Given('a {string} bucket', async function (versioning) {
+Given('an existing bucket {string} {string} versioning', async function (bucketName, hasVersioning) {
+    this.resetCommand();
+    const preName = this.parameters.AccountName || Constants.ACCOUNT_NAME;
+    if (bucketName === '') {
+        bucketName = `${preName}${Constants.BUCKET_NAME_TEST}${Utils.randomString()}`.toLocaleLowerCase();
+    }
```
```suggestion
    bucketName = bucketName || `${preName}${Constants.BUCKET_NAME_TEST}${Utils.randomString()}`.toLocaleLowerCase();
```

XinLi0207 commented on tests/ctst/common/common.ts:
```diff
@@ -15,3 +17,33 @@ Given('a {string} bucket', async function (versioning) {
+        }
+        await S3.putObject(this.getCommandParameters());
+    }
+});
```
let's add the last empty line

XinLi0207 commented on tests/ctst/steps/iam-policies/common.ts:
```diff
@@ -0,0 +1,217 @@
+    } else {
+        userCredentials = this.parameters.AssumedSession;
+    }
+    const api = testAPIs.find(api => api.API === action);
```
let's raise an error if api is not found in the list

KazToozs replied:
https://github.com/scality/Zenko/pull/1809/files#diff-284f11d6681baa8186a18ff40a18801ef2f353d3d23a6d8be5d52e3367ed92c9R180

this ok?

XinLi0207 commented on tests/ctst/world/Zenko.ts:
```diff
@@ -53,19 +60,21 @@ export default class Zenko extends World {
             }
-            this.cliMode.parameters.AssumedSession = CacheHelper.ARWWI[CacheHelper.AccountName];
-            this.cliMode.assumed = true;
+            // this.cliMode.parameters.AssumedSession = CacheHelper.ARWWI[CacheHelper.AccountName];
```
we can remove the comments

williamlardier commented on tests/ctst/world/Zenko.ts:
```diff
@@ -99,6 +106,326 @@ export default class Zenko extends World {
+            // Create credentials for the user
+            this.addCommandParameter({ userName: Zenko.IAMUserName });
+            const accessKey = await IAM.createAccessKey(this.getCommandParameters());
+            this.parameters.IAMSession = JSON.parse(accessKey.stdout).AccessKey;
```
Do we want to properly handle errors here? If the previous API call fails for some reason, the error message might be misleading here (and in IAM)

williamlardier commented on tests/ctst/common/common.ts:
```diff
@@ -15,3 +17,31 @@ Given('a {string} bucket', async function (versioning) {
+        if (withVersioning === 'with') {
+            this.saved.versionId = '3sL4kqtJlcpXroDTDmJ+rmSpXd3dIbrHY+MTRCxf3vjVBH40Nr8X8gdRQBpUMLUo';
+            this.addCommandParameter({versionId: this.saved.versionId});
+        }
```
Should be removed, we cannot specify a version ID when putting an object, this is done by the backend. If the bucket is version-enabled, it will be returned by the first command.

williamlardier replied:
Closing as resoved in latest commit.

williamlardier commented on .github/scripts/end2end/run-e2e-ctst.sh:
```diff
@@ -9,24 +9,41 @@ RETRIES=${4:-3}
 
 # Setting CTST world params
-WORLD_PARAMETERS='{"subdomain":"'${SUBDOMAIN}'","ssl":false,"port":"'${ZENKO_PORT}'","AccountName":"'${ZENKO_ACCOUNT_NAME}'","AccountAccessKey":"'${ZENKO_ACCESS_KEY}'","AccountSecretKey":"'${ZENKO_SECRET_KEY}'","AccountSessionToken":"'${ZENKO_SESSION_TOKEN}'","NotificationDestination":"'${NOTIF_DEST_NAME}'","NotificationDestinationTopic":"'${NOTIF_DEST_TOPIC}'","NotificationDestinationAlt":"'${NO ...
+WORLD_PARAMETERS='{"subdomain":"'${SUBDOMAIN}'","ssl":false,"port":"'${ZENKO_PORT}'","AccountName":"'${ZENKO_ACCOUNT_NAME}'","AdminAccessKey":"'${ADMIN_ACCESS_KEY_ID}'","AdminSecretKey":"'${ADMIN_SECRET_ACCESS_KEY}'","NotificationDestination":"'${NOTIF_DEST_NAME}'","NotificationDestinationTopic":"'${NOTIF_DEST_TOPIC}'","NotificationDestinationAlt":"'${NOTIF_ALT_DEST_NAME}'","NotificationDestinati ...
```
Just a formatting comment, given the number of parameters we could better display this, to make the review/edit process easy, what do you think? (not a blocker here)

KazToozs replied:
f87c3c881062fc60e0b746ebce2e0cc542345dc5

KazToozs replied:
ec9c50025db03abec6891bab3547478d580353e8

initial solution was for JS, not shell, oops

KazToozs replied:
Could not find reasonable solution for this, resolving as not blocking

williamlardier commented on tests/ctst/common/common.ts:
```diff
@@ -15,3 +17,31 @@ Given('a {string} bucket', async function (versioning) {
     }
 });
+
+Given('an existing bucket {string} {string} versioning', async function (bucketName, hasVersioning) {
```
Do we want a bit more options in this base scenario? Like the following parameters:

* Object lock enabled
* If yes, if we enable or not the retention
* If retention is enabled, whether we want it in governance or compliance mode.

This way, we can better reuse this common step definition. Another alternative would be to create other steps for these scenarios, but it duplicates the code.

KazToozs replied:
22d2f30b793b72bbd174ea27a9693b1e6900d79e

this look ok?

williamlardier commented on .github/scripts/end2end/run-e2e-ctst.sh:
```diff
@@ -36,4 +53,5 @@ kubectl run $POD_NAME \
         --image-pull-policy=Always \
         --env=TARGET_VERSION=$VERSION  \
-        -- ./run "$COMMAND" $WORLD_PARAMETERS "--parallel $PARALLEL_RUNS --retry $RETRIES --retry-tag-filter @Flaky"
+        --env=VERBOSE=1 \
```
Note for other reviewers: I think we should keep this flag set in the CI, t ease debugging in case of error, as this flag will display all the aws cli commands issued to Zenko, with all details about the responses.

williamlardier commented on .github/actions/deploy/action.yaml:
```diff
@@ -52,7 +52,7 @@ runs:
       run: bash deploy-zenko.sh end2end default
       working-directory: ./.github/scripts/end2end
-    - name: Add Keycloak user
+    - name: Add Keycloak user and assign StorageManager role
```
Picking random line:

It would be great to create a markdown page that explains how to contribute to the CTST tests in Zenko CI, so that we can also explain how they are supposed to work, i.e.:

- The tests are idempotent: they must not conflict with each other, here by using random bucket/object names.
- It should be possible to rerun the tests after the end of the CTST test suite (so, tests are not dependent on the environment).
And then maybe some other rules to follow, what do you think?

KazToozs replied:
a1872b532f5fd6ea0924a55df3623d859e43baba

thoughts?

williamlardier replied:
> Beyond usual Cucumber.js practice of using worlds, steps and features, this test suite has some specifics to consider:

This is something all tests should have in common, this is not specific to CTST, but an important reminder yes!

You could also add that CTST comes along with the AWS-CLI tool and abstraction layers to run it from TS code, but can be extended if needed (iwht iether cli tools or pure http tools, like axios).

Commit:
ZENKO-4499: improve Zenko world
Add user types and their access keys for use with CLI tests.
Add all expected S3 responsecode functions.

Commit:
ZENKO-4499: add common iam policy CTST functions

Commit:
ZENKO-4499: update keycloak config

Commit:
ZENKO-4499: update e2e environment and project setup files

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-4499 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.6.11`


Please check the `Fix Version/s` of ZENKO-4499, or the target
branch of this pull request.

KazToozs commented on .github/scripts/end2end/run-e2e-ctst.sh:
```diff
@@ -9,24 +9,41 @@ RETRIES=${4:-3}
+docker run \
+    --rm \
+    --network=host \
+    registry.scality.com/cli-testing-dev/cli-testing:"${CTST_VERSION}" /bin/bash \
```
E2E_IMAGE comes out as `registry.scality.com/zenko/zenko-e2e-ctst` from the end2end.yaml file, so for the moment we don't have an appropriate env variable. Any other suggestion here?

XinLi0207 commented on .github/scripts/end2end/configs/keycloak_config.json:
```diff
@@ -19,7 +19,7 @@
     }
   ],
-  "defaultRoles" : [ "uma_authorization", "StorageManager", "offline_access" ],
+  "defaultRoles" : [ "uma_authorization", "offline_access" ],
```
leave a comment for reviewers
`StorageManager` should not be set as defaultRole for keycloak users, but it didn't affect existing keycloak users, because existing keycloak users overwrite their roles. but it does affect new added keycloak users, so need to fix here

Commit:
ZENKO-4499: post comment: handle createAccessKey error

anurag4DSB commented on tests/ctst/world/Zenko.ts:
```diff
@@ -35,11 +36,17 @@ export default class Zenko extends World {
-    private saved: object = {};
+    private saved: any = {};
+
+    private static IAMUserName = '';
```
(adding in one place but multiple places have same occurrence)
why aren't we declaring types here?
Example this can also be written as 

```js
private static IAMUserName: string = '';
```

KazToozs replied:
TypeScript infers a type at initialisation
https://www.tutorialsteacher.com/typescript/type-inference#:~:text=TypeScript%20infers%20types%20of%20variables,values%20are%20set%20for%20parameters

But I'm open to making changes to variables you feel are worth typing explicitly if you insist on this

anurag4DSB commented on tests/ctst/world/Zenko.ts:
```diff
@@ -99,6 +106,318 @@ export default class Zenko extends World {
+        const savedParameters = JSON.parse(JSON.stringify(this.cliOptions));
+        this.resetGlobalType();
+        switch (type) {
+            case EntityType.ACCOUNT:
+                await this.prepareRootUser();
+                this.saved.type = EntityType.IAM_USER;
+                break;
+            case EntityType.IAM_USER:
+                await this.prepareIamUser();
+                this.saved.type = EntityType.ACCOUNT;
```
Why are these flipped?
for `case EntityType.ACCOUNT` we are using `this.saved.type = EntityType.IAM_USER;`
and vice versa

XinLi0207 replied:
😅  will switch them back, thanks for catching

KazToozs replied:
669a8fddbdcf3a3d1f57ee3645ade64ad2a03996

anurag4DSB commented on tests/ctst/world/Zenko.ts:
```diff
@@ -99,6 +106,318 @@ export default class Zenko extends World {
+     * 'storage account owner', 'data consumer' or 'iam user'
+     * @returns {undefined}
+     */
+    async prepareForType(type: string): Promise<void> {
```
nit: worth rewording to make it more intutive
`prepareForType` -> `setupEntity`
`type` parameter -> `entityType`

KazToozs replied:
ef214455bb346bf5cde53f71ac79e06eca513be3

anurag4DSB commented on tests/ctst/world/Zenko.ts:
```diff
@@ -99,6 +106,318 @@ export default class Zenko extends World {
+                name: this.parameters.AccountName || Constants.ACCOUNT_NAME,
+            });
+            // Getting roles with GetRolesForWebIdentity
+            const roles = await SuperAdmin.getRolesForWebIdentity(this.options.webIdentityToken);
```
nit: worth destructuring for readability and reduce repeatability later on, for roles.data.data
```
const { data { data: roles } } = await SuperAdmin.getRolesForWebIdentity(this.options.webIdentityToken);
```

KazToozs replied:
2491f6211da707d8a7ad043996d61ee10e8fbbf8

anurag4DSB commented on tests/ctst/world/Zenko.ts:
```diff
@@ -99,6 +106,318 @@ export default class Zenko extends World {
+            },
+            data,
+        };
+        const result: AxiosResponse = await axios(config);
+        return result.data.access_token;
```
Do we not care about any error handling here, considering we are making an API request for returning an access_token and we possibly have many functions relying on it?

PS: Not reviewed how it is used yet.

williamlardier replied:
Usually the approach for tests, from my experience, is "let it throw", as it will return an error anyway...

It may be better to throw a known and human-readable error though, to ease the debugging, as long as enough information is logged along with the message

anurag4DSB replied:
yeah! aiming for easy debugging as well.
If you think error is clear enough, we should be good.
Feel free to resolve the comment if that's the case.

KazToozs replied:
455f9416a39b4307accf57abb26b7b87a86b3662

thoughts?

williamlardier replied:
perfect!

KazToozs commented on tests/ctst/steps/iam-policies/common.ts:
```diff
@@ -0,0 +1,218 @@
+    }
+    const api = testAPIs.find(api => api.API === action);
+    assert(api);
+    if (ifS3Standard === 'notS3Standard'){
```
3c6d7c63e9e9d6f3ef640927915e92067e49a2a9

This seem ok?

KazToozs replied:
https://github.com/scality/Zenko/pull/1809/commits/bcb2e258ea0b45c91357224397bf7796eab99f1c#diff-284f11d6681baa8186a18ff40a18801ef2f353d3d23a6d8be5d52e3367ed92c9R15

ok?

williamlardier replied:
All god, the scope inside `{ ... }` and the last `break` are not formally needed, but resolving the comment, it's a minor suggestion

anurag4DSB commented on tests/ctst/world/Zenko.ts:
```diff
@@ -99,6 +106,318 @@ export default class Zenko extends World {
+     * Erases all the environment configuration from prepareForType
+     * @returns {undefined}
+     */
+    endForType(): void {
```
nit: worth rewording to use the keyword `cleanup` as a part of the name of the function to make it more intuitive 
adding to the comment above where I suggested rewording the `prepareForType` function

KazToozs replied:
fa26d853481674bb9c2b12689a0f06340526a4cb

anurag4DSB replied:
added one more comment: https://github.com/scality/Zenko/commit/fa26d853481674bb9c2b12689a0f06340526a4cb#r102400941

KazToozs replied:
Addressed https://github.com/scality/Zenko/commit/fa26d853481674bb9c2b12689a0f06340526a4cb#r102419022

KazToozs commented on tests/ctst/world/Zenko.ts:
```diff
@@ -99,6 +106,326 @@ export default class Zenko extends World {
+            case EntityType.STORAGE_MANAGER:
+                await this.prepareARWWI(
+                    this.parameters.StorageManagerUsername || 'storage_manager',
+                    this.parameters.KeycloakTestPassword || '123',
```
d88ee6e72f4271d89dd51cfa6a48de5d20df8ce5

anurag4DSB commented on tests/ctst/world/Zenko.ts:
```diff
@@ -99,6 +106,318 @@ export default class Zenko extends World {
+    }
+
     /**
      * Helper function to change the default aws cli command type (iam|sts|s3).
```
might be out of scope, but here we are using a service, and `type` for service, hence a bit confusing.

KazToozs replied:
70220c200e70dad2d88d35cd04696e8c4e3c8cc1

No found occurrences in the code of this function and it's variable being used, so changes were minimal

anurag4DSB commented on tests/ctst/common/utils.ts:
```diff
@@ -0,0 +1,74 @@
+}
+
+export const s3FunctionExtraParams : { [key: string]: Object } = {
+    'restoreObject': { restoreRequest: 'Days=1' },
```
we did merge recently Days can be zero in this PR: https://github.com/scality/pensieve-api/pull/242
Not sure if its usable right away as it might depend on releases.

XinLi0207 replied:
that's good to know. Actually in our cases we don't care about valid or invalid values for specific parameters, we are testing permission check, as long as we provide any valid parameter and the request don't get `wrong parameter error`  we should be good.
so for some specific parameters, we hardcoded them

anurag4DSB commented on .github/scripts/end2end/keycloak-helper.sh:
```diff
@@ -37,6 +37,12 @@ case $COMMAND in
+        # attach StorageManager role to user
+        ${KEYCLOAK_EXEC} /opt/jboss/keycloak/bin/kcadm.sh add-roles \
+          -r ${OIDC_REALM} \
+          --uusername ${OIDC_USERNAME} \
```
Is this a typo, there are 2 "u" in username

XinLi0207 replied:
there are 2 u
this is how keycloak defined this command
https://github.com/keycloak/keycloak-documentation/blob/main/server_admin/topics/admin-cli.adoc#adding-client-roles-to-a-user

anurag4DSB replied:
Thanks for educating me :)

anurag4DSB reviewed (commented):
Iteration 1 of review

anurag4DSB reviewed (approved):
LGTM, we already have a document in progress for the way to add tests, which is good.
Added one small comment on `this.cleanupEntity()` method.

Reference: https://github.com/scality/Zenko/pull/1809#discussion_r1114298765

Commit:
ZENKO-4499: refactor AWS 'type' to 'service'

Commit:
ZENKO-4499: add contributing guidelines

KazToozs commented on .github/scripts/end2end/run-e2e-ctst.sh:
```diff
@@ -36,4 +53,5 @@ kubectl run $POD_NAME \
         --env=TARGET_VERSION=$VERSION  \
-        -- ./run "$COMMAND" $WORLD_PARAMETERS "--parallel $PARALLEL_RUNS --retry $RETRIES --retry-tag-filter @Flaky"
+        --env=VERBOSE=1 \
+        -- ./run "$COMMAND" $WORLD_PARAMETERS "--parallel $PARALLEL_RUNS --retry $RETRIES --retry-tag-filter @Flaky"
```
a1872b532f5fd6ea0924a55df3623d859e43baba

thoughts?

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-4499 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.6.12`


Please check the `Fix Version/s` of ZENKO-4499, or the target
branch of this pull request.

williamlardier reviewed (approved):
LGTM, beyond the last minor comments

XinLi0207 commented on .github/workflows/end2end.yaml:
```diff
@@ -555,6 +556,8 @@ jobs:
       - name: Run CTST end to end tests
         run: bash run-e2e-ctst.sh
         working-directory: ./.github/scripts/end2end
+      - name: Setup tmate session  
```
let's remove the tmate session

KazToozs commented:
/approve

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-4499 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.6.12`


Please check the `Fix Version/s` of ZENKO-4499, or the target
branch of this pull request.




*The following options are set:* **approve**

KazToozs commented:
ping

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue ZENKO-4499.

Goodbye kaztoozs.
